lib/src/default_index/revset_engine.rs

```diff
@@ -130,8 +130,13 @@ impl<I: AsCompositeIndex> Revset for RevsetImpl<I> {
         self.entries().next().is_none()
     }
 
-    fn count(&self) -> usize {
-        self.entries().count()
+    fn count_estimate(&self) -> (usize, Option<usize>) {
+        let count = self.entries().take(50).count();
+        if count < 50 {
+            (count, Some(count))
+        } else {
+            (50, None)
+        }
```

Comment:
On local disk (which default_index is), it's pretty cheap to walk relatively large revsets. I think it's fine even in the somewhat rare case of a branch that's tens of thousands ahead/behind its remote version. If we later replace the current default_index by something based on segmented changelog, then we can usually give an exact count even faster, even for large revsets. So I would prefer to keep the old implementation here and just let Google use the estimate where it's expensive to give an exact number. What do you think?

Reply:
Even on local disk, GitHub found the need to improve ahead/behind checking in Git, as you can see here: https://lore.kernel.org/git/pull.1489.git.1678111598.gitgitgadget@gmail.com/

Among other things, they implemented a mode that only checked "zero commits ahead" (versus more than zero commits ahead) and a new algorithm that requires generation numbers. There are reasons to say that jj is different from Git here (since the starting point for Git - one rev-list invocation per thing we want to check - is different from jj's starting point, and GitHub's scale is different) but I still think that we should optimize it.

Besides the above, one logistical reason for teaching the default_index's Revset to return estimates is to avoid untested code in jj. If default_index's Revset only ever returned exact values, then we wouldn't be able to test branch's output when the Revset implementation decides to return an estimate. (The alternative, to not teach branch anything and instead make it call iter().count() doesn't work for backends that have expensive commit reading, like Google's case.) This logistical reason shouldn't be the only reason why we do things, but perhaps it could be a tiebreaker if people are ambivalent about this change.

Reply:
We always use generation numbers and don't read objects when doing revision walks, so maybe it's no so bad? It seems to add 170 ms when one of my commits was 1.1M commits behind in the Linux kernel. Anyway, I'm also fine with iterating a fixed number of commits because I don't think anyone really care if it's 1.1M or 1.0M commits (and hopefully not if it's 1.1M or 100 either).

I agree that it's nice to be able to test it.

Reply:
OK, using your suggestion regarding test-fakes (as described in another thread), we won't use estimation in production but we can still test it, so my concerns are resolved.